lightning_pose/models/heatmap_tracker_mhcrnn.py

```diff
@@ -0,0 +1,421 @@
+            output_shape=output_shape,
+            torch_seed=torch_seed,
+            lr_scheduler=lr_scheduler,
+            lr_scheduler_params=lr_scheduler_params,
+        )
+        self.loss_factory_unsup = loss_factory_unsupervised.to(self.device)
+
+        # this attribute will be modified by AnnealWeight callback during training
+        # self.register_buffer("total_unsupervised_importance", torch.tensor(1.0))
+        self.total_unsupervised_importance = torch.tensor(1.0)
+
+    def get_loss_inputs_unlabeled(self, batch: UnlabeledBatchDict) -> dict:
+        """Return predicted heatmaps and their softmaxes (estimated keypoints)."""
+        # images -> heatmaps
+        pred_heatmaps_crnn, pred_heatmaps_sf = self.forward(batch["frames"])
+        # heatmaps -> keypoints
+        pred_keypoints_crnn, confidence_crnn = self.run_subpixelmaxima(pred_heatmaps_crnn)
+        pred_keypoints_sf, confidence_sf = self.run_subpixelmaxima(pred_heatmaps_sf)
+
+        # undo augmentation if needed
+        if batch["transforms"].shape[-1] == 3:
+            # reshape to (seq_len, n_keypoints, 2)
+            pred_kps = torch.reshape(pred_keypoints_crnn, (pred_keypoints_crnn.shape[0], -1, 2))
+            # undo
+            pred_kps = undo_affine_transform(pred_kps, batch["transforms"])
+            # reshape to (seq_len, n_keypoints * 2)
+            pred_keypoints_crnn = torch.reshape(pred_kps, (pred_kps.shape[0], -1))
+
+            # reshape to (seq_len, n_keypoints, 2)
+            pred_kps = torch.reshape(pred_keypoints_sf, (pred_keypoints_sf.shape[0], -1, 2))
+            # undo
+            pred_kps = undo_affine_transform(pred_kps, batch["transforms"])
+            # reshape to (seq_len, n_keypoints * 2)
+            pred_keypoints_sf = torch.reshape(pred_kps, (pred_kps.shape[0], -1))
+
+        return {
+            "heatmaps_pred": torch.cat([pred_heatmaps_crnn, pred_heatmaps_sf], dim=0),
+            "keypoints_pred": torch.cat([pred_keypoints_crnn, pred_keypoints_sf], dim=0),
+            "confidences": torch.cat([confidence_crnn, confidence_sf], dim=0),
+        }
+
+    def get_parameters(self):
+        params = [
+            # don't uncomment line below
+            # the BackboneFinetuning callback should add backbone to the params.
+            # {"params": self.backbone.parameters()},
+            # important this is the 0th element, for BackboneFinetuning callback
+            {"params": self.upsampling_layers_rnn.parameters()},
+            {"params": self.upsampling_layers_sf.parameters()},
+        ]
+        # define different learning rate for weights in front of unsupervised losses
+        if len(self.loss_factory_unsup.loss_weights_parameter_dict) > 0:
+            params.append({
+                "params": self.loss_factory_unsup.loss_weights_parameter_dict.parameters(),
+                "lr": 1e-2,
+            })
+        return params
+
+
+class UpsamplingCRNN(torch.nn.Module):
+
+    def __init__(
```

Comment:
more typechecking?
might be good if we add a formula in a comment that links those W's and H's.

Reply:
added typechecking, we can ask cole to add a formula later